File: 1207-delete-nodes-and-return-forest/1207-delete-nodes-and-return-forest.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def dfs(self, root, st, ans):
        if not root:
            return None
        root.left, root.right = self.dfs(root.left, st, ans), self.dfs(root.right, st, ans)
        
        if root.val in st:
            if root.left: ans.append(root.left)
            if root.right: ans.append(root.right)
            return None
        
        return root

    def delNodes(self, root, to_delete):
        ans = []
        st = set(to_delete)
        root = self.dfs(root, st, ans)
        if root: ans.append(root)

        return ans
        

File: 1207-delete-nodes-and-return-forest/1207-delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode dfs(TreeNode root, HashSet<Integer> st, List<TreeNode> ans) {
        if(root == null) return root;
        TreeNode l = dfs(root.left, st, ans);
        TreeNode r = dfs(root.right, st, ans);

        root.left = l;
        root.right = r;

        if(st.contains(root.val)) {
            if(root.left != null) ans.add(root.left);
            if(root.right != null) ans.add(root.right);

            return null;
        }

        return root;
    }

    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        List<TreeNode> ans = new ArrayList<>();
        HashSet<Integer> st = new HashSet<>();
        for(int n : to_delete) st.add(n);

        root = dfs(root, st, ans);
        if(root != null) ans.add(root);

        return ans;
    }
}